src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1947,18 +1949,32 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
             }
 
             Rvalue::Repeat(operand, len) => if *len > 1 {
-                let operand_ty = operand.ty(mir, tcx);
-
-                let trait_ref = ty::TraitRef {
-                    def_id: tcx.lang_items().copy_trait().unwrap(),
-                    substs: tcx.mk_substs_trait(operand_ty, &[]),
-                };
-
-                self.prove_trait_ref(
-                    trait_ref,
-                    location.to_locations(),
-                    ConstraintCategory::CopyBound,
-                );
+                if let Operand::Move(_) = operand {
+                    // While this is located in `nll::typeck` this error is not an NLL error, it's
+                    // a required check to make sure that repeated elements implement `Copy`.
+                    let span = mir.source_info(location).span;
+                    let ty = operand.ty(mir, self.tcx());
+                    let is_copy = self.infcx.type_is_copy_modulo_regions(self.param_env, ty, span);
+                    if !is_copy {
```

Comment:
Can you move the diagnostics logic to a separate function to keep the happy path cleaner?

Reply:
I've started re-using some existing functions for diagnostics here.